File: src/Wysicraft.Core/ResponsiveLayout.cs
using Wysicraft.Models;
namespace Wysicraft.Core;

public static class ResponsiveLayout
{
    public static Dictionary<string,Bounds> Resolve(UiDefinition design,double width,double height)=>Layout(design,width,height,null,null);
    // Editor reflow: the bounds of every descendant of `id` after that element is resized from its design bounds to `bounds`.
    public static Dictionary<string,Bounds> Reflow(UiDefinition design,string id,Bounds bounds) {
        var all=Layout(design,design.Size.Width,design.Size.Height,id,bounds);
        return all.Where(p=>IsDescendant(design,p.Key,id)).ToDictionary(p=>p.Key,p=>p.Value);
    }
    static bool IsDescendant(UiDefinition design,string candidate,string ancestor) {
        var parent=design.Elements.FirstOrDefault(e=>e.Id==candidate)?.Parent;
        for(int depth=0;!string.IsNullOrEmpty(parent) && depth<=design.Elements.Count;depth++) {
            if(parent==ancestor)return true;
            parent=design.Elements.FirstOrDefault(e=>e.Id==parent)?.Parent;
        }
        return false;
    }
    static Dictionary<string,Bounds> Layout(UiDefinition design,double width,double height,string? fixedId,Bounds? fixedBounds) {
        var resolved=new Dictionary<string,Bounds>();var visiting=new HashSet<string>();
        Bounds Place(Element e) {
            if(resolved.TryGetValue(e.Id,out var found))return found;
            if(e.Id==fixedId && fixedBounds!=null)return resolved[e.Id]=Json.Clone(fixedBounds);
            if(!visiting.Add(e.Id))throw new InvalidDataException("Container cycle");
            var parent=design.Elements.FirstOrDefault(p=>p.Id==e.Parent);
            var old=parent?.Bounds ?? new Bounds {Width=design.Size.Width,Height=design.Size.Height};
            var next=parent==null?new Bounds {Width=width,Height=height}:Place(parent);
            double dw=next.Width-old.Width,dh=next.Height-old.Height;
            var b=Json.Clone(e.Bounds);b.X+=next.X-old.X;b.Y+=next.Y-old.Y;
            switch(e.HorizontalAnchor) {case "right":b.X+=dw;break;case "center":b.X+=dw/2;break;case "stretch":b.Width=Math.Max(e.MinWidth,b.Width+dw);break;}
            switch(e.VerticalAnchor) {case "bottom":b.Y+=dh;break;case "center":b.Y+=dh/2;break;case "stretch":b.Height=Math.Max(e.MinHeight,b.Height+dh);break;}
            visiting.Remove(e.Id);return resolved[e.Id]=b;
        }
        foreach(var e in design.Elements)Place(e);return resolved;
    }
    public static void Apply(UiDefinition target,UiDefinition design,int width,int height) {
        width=design.Responsive?Math.Clamp(width,16,4096):design.Size.Width;
        height=design.Responsive?Math.Clamp(height,16,4096):design.Size.Height;
        var bounds=Resolve(design,width,height);
        foreach(var e in target.Elements)e.Bounds=Json.Clone(bounds[e.Id]);
        target.Size=new(){Width=width,Height=height};
    }
}

File: wysicraft-runtime/src/main/java/com/wysicraft/runtime/model/ResponsiveLayout.java
package com.wysicraft.runtime.model;
import com.wysicraft.runtime.model.Models.*;
import java.util.*;

public final class ResponsiveLayout {
    public static Map<String,Bounds> resolve(Ui design,double width,double height) {
        var result=new HashMap<String,Bounds>();var visiting=new HashSet<String>();
        class Resolver {
            Bounds place(Element e) {
                if(result.containsKey(e.id))return result.get(e.id);
                if(!visiting.add(e.id))throw new IllegalArgumentException("Container cycle");
                var parent=design.element(e.parent);var old=new Bounds();old.width=design.size.width;old.height=design.size.height;
                var next=new Bounds();next.width=width;next.height=height;
                if(parent!=null){old=parent.bounds;next=place(parent);}
                double dw=next.width-old.width,dh=next.height-old.height;
                var b=new Bounds();b.x=e.bounds.x+next.x-old.x;b.y=e.bounds.y+next.y-old.y;b.width=e.bounds.width;b.height=e.bounds.height;
                switch(e.horizontalAnchor){case "right"->b.x+=dw;case "center"->b.x+=dw/2;case "stretch"->b.width=Math.max(e.minWidth,b.width+dw);}
                switch(e.verticalAnchor){case "bottom"->b.y+=dh;case "center"->b.y+=dh/2;case "stretch"->b.height=Math.max(e.minHeight,b.height+dh);}
                visiting.remove(e.id);result.put(e.id,b);return b;
            }
        }
        var resolver=new Resolver();for(var e:design.elements)resolver.place(e);return result;
    }
    public static void apply(Ui target,Ui design,int width,int height) {
        width=design.responsive?Math.clamp(width,16,4096):design.size.width;height=design.responsive?Math.clamp(height,16,4096):design.size.height;
        var bounds=resolve(design,width,height);
        for(var e:target.elements) {
            e.bounds=bounds.get(e.id);var source=design.element(e.id);
            if(!source.rowElements.isEmpty()) {
                var row=Models.JSON.fromJson(Models.JSON.toJson(source),Element.class);var layout=RowTemplates.layout(source);
                layout.size.width=(int)(source.rowTemplateWidth>0?source.rowTemplateWidth:source.bounds.width);layout.size.height=source.rowHeight;
                var rowBounds=resolve(layout,e.bounds.width,e.rowHeight);
                for(var child:row.rowElements)child.bounds=rowBounds.get(child.id);e.rowElements=row.rowElements;
            }
        }
        target.size.width=width;target.size.height=height;
    }
}
